src/alert.cpp

```diff
@@ -1829,5 +1829,31 @@ namespace libtorrent {
 		}
 	}
 
+	dht_direct_response_alert::dht_direct_response_alert(
+		aux::stack_allocator& alloc, void* userdata
+		, udp::endpoint const& addr, bdecode_node const& response)
+		: userdata(userdata), addr(addr), m_alloc(alloc)
+		, m_response_idx(alloc.copy_buffer(response.data_section().first, response.data_section().second))
+		, m_response_size(response.data_section().second)
+	{}
+
+	std::string dht_direct_response_alert::message() const
+	{
+		char msg[1050];
+		snprintf(msg, sizeof(msg), "DHT direct response (address=%s) [ %s ]"
+			, addr.address().to_string().c_str()
+			, std::string(m_alloc.ptr(m_response_idx), m_response_size).c_str());
+		return msg;
+	}
+
+	void dht_direct_response_alert::response(bdecode_node& ret) const
```

Comment:
you could even return the bdecode_node here. I suppose the internal parse "tree" may be copied, but on a modern compiler it should be moved. The underlying buffer won't get copied either way.